doc/inband-telemetry/SAI-Proposal-Advanced-Telemetry-draft.md

```diff
@@ -0,0 +1,421 @@
+~~~cpp
+/**
+ * @brief Telemetry feature types
+ * INT_EP, INT_TRANSIT, POSTCARD are mutually exclusive
+ * MOD can coexist with any one of the three above
+ */
+typedef enum _sai_telemetry_type_t {
+    /** INT source and sink */
+    SAI_TELEMETRY_TYPE_INT_EP,
+    /** INT transit */
+    SAI_TELEMETRY_TYPE_INT_TRANSIT,
+    /** Postcard */
+    SAI_TELEMETRY_TYPE_POSTCARD,
+    /** Mirror on Drop */
+    SAI_TELEMETRY_TYPE_MOD,
+} sai_telemetry_type_t;
+
+/**
+ * @brief Telemetry attributes
+ */
+typedef enum _sai_telemetry_attr_t {
+    /** List of telemetry features to enable */
+    SAI_TELEMETRY_ATTR_TYPE_LIST,
+    /** Globally unique switch ID */
+    SAI_TELEMETRY_ATTR_SWITCH_ID,
+    /** List of ERSPAN mirror sessions for report sending*/
+    SAI_TELEMETRY_ATTR_MIRROR_LIST,
+
+    /** Reserved DSCP value for INT */
+    SAI_TELEMETRY_ATTR_INT_DSCP,
+    /** INT sink downstream ports */
+    SAI_TELEMETRY_ATTR_INT_SINK_PORT_LIST,
+} sai_telemetry_attr_t;
+~~~
+
+#### Telemetry Watchlist ####
+~~~cpp
+/** Telemetry watchlist match field enum */
+typedef enum _sai_telemetry_watchlist_entry_attr_t {
+    /** Telemetry type, each type has its own watchlist table */
+    SAI_TWL_ENTRY_ATTR_TELEMETRY_TYPE,
+    /** Entry priority, only compared within the same telemetry type */
+    SAI_TWL_ENTRY_ATTR_PRIORITY,
+
+    /** Ethernet type */
+    SAI_TWL_ENTRY_ATTR_FIELD_ETHER_TYPE,
+    /** IPv4 source address */
+    SAI_TWL_ENTRY_ATTR_FIELD_IPV4_SRC,
+    /** IPv4 destination address */
+    SAI_TWL_ENTRY_ATTR_FIELD_IPV4_DST,
+    /** IP protocol */
+    SAI_TWL_ENTRY_ATTR_FIELD_IP_PROTO,
+    /** L4 source port UDP/TCP */
+    SAI_TWL_ENTRY_ATTR_FIELD_L4_PORT_SRC,
+    /** L4 destination port UDP/TCP */
+    SAI_TWL_ENTRY_ATTR_FIELD_L4_PORT_DST,
+    /** Tunnel vni */
+    SAI_TWL_ENTRY_ATTR_FIELD_TUNNEL_VNI,
+    /** Ethernet type */
+    SAI_TWL_ENTRY_ATTR_FIELD_INNER_ETHER_TYPE,
+    /** inner IPv4 Source address */
+    SAI_TWL_ENTRY_ATTR_FIELD_INNER_IPV4_SRC,
+    /** inner IPv4 Destination address */
+    SAI_TWL_ENTRY_ATTR_FIELD_INNER_IPV4_DST,
+    /** inner IP Protocol */
+    SAI_TWL_ENTRY_ATTR_FIELD_INNER_IP_PROTO,
+    /** L4 source port UDP/TCP */
+    SAI_TWL_ENTRY_ATTR_FIELD_INNER_L4_PORT_SRC,
+    /** L4 destination port UDP/TCP */
+    SAI_TWL_ENTRY_ATTR_FIELD_INNER_L4_PORT_DST,
+
+    /** Watch or not */
+    SAI_TWL_ENTRY_ATTR_ACTION_WATCH,
+    /** Report every matched packet (no triggering event detection) */
+    SAI_TWL_ENTRY_ATTR_ACTION_REPORT_ALL,
+    /** INT config session ID */
+    SAI_TWL_ENTRY_ATTR_ACTION_INT_CONFIG_SESSION_ID,
+} sai_telemetry_watchlist_entry_attr_t;
+~~~
+
+#### Telemetry report triggering event detection ####
+~~~cpp
+/** Queue alert thresholds */
+typedef struct _sai_telemetry_queue_alert_t {
+    /** egress port */
+    uint16_t port;
+    /** queue id */
+    uint16_t queue;
+    uint32_t depth_threshold;
+    uint32_t latency_threshold;
+} sai_telemetry_queue_alert_t;
```

Comment:
sai_telemetry_queue_alert_t seem to extra entry in sai_attribute_value_t
and this contains port and queue, should this be object id's instead of uint's ? if yes please convert this sai_telemetry_queue_alert_t to as new SAI_OBJECT_TYPE_TELEMETRY_QUEUE_ALERT, and all those fileds mark as attirbutes of that object, and mark SAI_TELEMETRY_TRIGGER_ATTR_QUEUE_ALERT as sai_object_id_t
since we don't want to have object_id fields insice attr_value structure other than oid and objlist since it will be very hard for metadata to track